### PR TITLE
[feature|bugfix] Add issue aliases - depends on #1

### PR DESCRIPTION
**Pull request by TimKolberger**
## 🥳 Issue aliases
Add a new issue alias by executing `juli alias my-daily-alias=ISSUE-123`
Remove an issue alias by executing `juli alias my-daily-alias=`

The aliases are saved in a separate `alias.json` file in the home directory.

## 🐞 Thursday bug
Fixes a bug which causes sync problems on thursdays

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -142,10 +142,10 @@ commander@^2.12.1:
   resolved "https://registry.yarnpkg.com/commander/-/commander-2.20.3.tgz#fd485e84c03eb4881c20722ba48035e8531aeb33"
   integrity sha512-GpVkmM8vF2vQUkj2LvZmD35JxeJOLCwJ9cUkugyk2nuhbv3+mJvpLYYt+0+USMxE+oj+ey/lJEnhZw75x/OMcQ==
 
-commander@^4.0.1:
-  version "4.0.1"
-  resolved "https://registry.yarnpkg.com/commander/-/commander-4.0.1.tgz#b67622721785993182e807f4883633e6401ba53c"
-  integrity sha512-IPF4ouhCP+qdlcmCedhxX4xiGBPyigb8v5NeUp+0LyhwLgxMqyp3S0vl7TAPfS/hiP7FC3caI/PB9lTmP8r1NA==
+commander@^4.1.0:
```
Comment (TimKolberger) [inline yarn.lock:202]: Please note that i updated commander to get async actions working

File: src/jira/JiraApiConnector.ts [hunk 2]
```diff
@@ -68,29 +97,40 @@ export class JiraApiConnector implements IApiConnector {
...
-    const preparedEntries = this.prepareEntries(days);
+    try {
+      await this.client.obtainCookie();
+    } catch (e) {
+      throw AuthenticationError.fromAxiosError(e);
+    }
 
+    const issueAliases = this.store.getIssueAliases();
+    const preparedEntries = this.prepareEntries(days, issueAliases);
+    console.log(JSON.stringify(preparedEntries, null, 2));
```
Comment (Agreon) [inline src/jira/JiraApiConnector.ts]: Please remove the console.log

File: src/jira/JiraClient.ts [hunk 3]
```diff
@@ -21,6 +21,10 @@ export interface IJiraWorklog {
   date: Date;
 }
 
+export interface IJiraAlias {
```
Comment (Agreon) [inline src/jira/JiraClient.ts]: I would move that to the `types.ts`-file as it is not a constraint from the api, but our own definition.

**Review by Agreon — CHANGES_REQUESTED** [review]
I would love to have the possibility to add an optional description to an alias like 'Daily Standup'. Do you think that this causes problems?

File: src/jira/JiraApiConnector.ts (hunk 2 above)
Comment (TimKolberger) [reply src/jira/JiraApiConnector.ts]: 👌

File: src/jira/JiraClient.ts (hunk 3 above)
Comment (TimKolberger) [reply src/jira/JiraClient.ts]: I extracted the alias management to a AliasStore

File: Readme.md [hunk 4]
```diff
@@ -40,7 +46,7 @@ $ juli <timesheet-file> [Params]
 - Format of the date: `d.M.[YY]`
 - Format of the the time: `h[:m]`
 - `: <DESCRIPTION>` can be omitted, if there is an existing issue-definition on the same day already.
-- `//` At the beginning of a line represent comments which will be ignored
+- `<ALIAS>` entries can be managed with the command `juli alias my-daily-alias=ISSUE-123`- `//` At the beginning of a line represent comments which will be ignored
```
Comment (Agreon) [inline Readme.md]: Führt zu:
Selection_153

File: src/Parser.ts [hunk 5]
```diff
@@ -1,10 +1,24 @@
 import { IWorkDay, IWorkEntry } from "./types";
 import { parseDate } from "./jira/util";
 import { FormatError } from "./errors";
+import { AliasStore } from "./repository/AliasStore";
 import { format } from "date-fns";
 
 export class Parser {
-  public static parse(text: string): IWorkDay[] {
+  private static parseTime(text: string): Date {
```
Comment (Agreon) [inline src/Parser.ts:8]: You can remove the static keyword now. It has no use case anymore

File: Readme.md [hunk 6]
```diff
@@ -64,6 +73,9 @@ Entries, that were added manually or through other tools are not affected by thi
 
 - `updateHost`: Update the saved Jira host.
 - `updateCredentials`: Update the saved credentials.
+- `alias [alias]`
+  - Upsert an alias with `<name>=[issueId],[comment]`
+  - Remove an alias with `<name>=`
```
Comment (Agreon) [inline Readme.md]: From the User/UX-Perspective i would not expect to give a '=' to delete an alias. Maybe you can resolve this by not putting the delete, update and create functionality into one method.

File: src/repository/AliasStore.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      await this.setIssueAliases(remainingAliases);
+      return remainingAliases;
+    }
+
+    const updatedAliases = { ...aliases, [name]: { ticketId, description } };
+    await this.setIssueAliases(updatedAliases);
+    return updatedAliases;
+  }
+
+  public resolveAlias: (idOrName: string) => IAlias = (idOrName: string) => {
```
Comment (Agreon) [inline src/repository/AliasStore.ts]: Why is this curried?

File: src/repository/AliasStore.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    const { name, ticketId, description } = matchObj.groups;
+    if (!name) {
+      throw new InvalidArgumentError(
+        `${alias} does not match <name>=<issueId>`
+      );
+    }
+
+    const aliases: IAliases = (await this.getIssueAliases()) || {};
```
Comment (Agreon) [inline src/repository/AliasStore.ts]: There is no need for await (or the method) anymore

File: src/index.ts [hunk 9]
```diff
@@ -28,6 +31,41 @@ export const execute = async () => {
...
+    .action(async (alias: string) => {
+      executed = true;
+
+      const printAliases = (aliases: IAliases | null) =>
+        aliases
+          ? Object.entries(
+              aliases
+            ).forEach(([name, { ticketId, description }]) =>
+              Logger.info(
+                `${name}=${ticketId}${description ? `,${description}` : ""}`
```
Comment (Agreon) [inline src/index.ts:46]: Could you put a ' ' after the ',' ?

File: src/repository/AliasStore.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
+import * as fs from "fs-extra";
+import { CREATE_PATH } from "../const";
+import * as path from "path";
+import { InvalidArgumentError } from "../errors/InvalidArgumentError";
+import { IAlias, IAliases } from "../types";
+
+export class AliasStore {
+  static readonly ALIAS_FILE = path.join("jira", "alias.json");
```
Comment (Agreon) [inline src/repository/AliasStore.ts]: Do you use the 'ALIAS_FILE'-Variable somewhere else or why is it in here? I would like to keep things consistent (like in JiraStore as consts) but if you got an good argument i'd also prefer your way :)

File: src/Parser.ts [hunk 11]
```diff
@@ -42,44 +56,55 @@ export class Parser {
 
     for (let i = 0; i < lines.length; i += 1) {
       if (!startTime) {
-        startTime = this.parseTime(lines[i]);
+        startTime = Parser.parseTime(lines[i]);
         continue;
       }
 
-      if (!description || !ticketId) {
+      if (!ticketId && !description && lines[i].includes(":")) {
```
Comment (Agreon) [inline src/Parser.ts:63]: Why did you add that constraint. That means, even if i got no description, i still have to add ':' after my ticket id. It worked without problems before...

File: src/Parser.ts [hunk 12]
```diff
@@ -42,44 +56,55 @@ export class Parser {
...
-
-      entries.push({
-        startTime,
-        endTime,
-        ticketId,
-        description
-      });
+      endTime = Parser.parseTime(lines[i]);
+
+      if (ticketId && description) {
```
Comment (Agreon) [inline src/Parser.ts:100]: If you remove the check for ':' you can also keep the structure down here.

**Review by Agreon — CHANGES_REQUESTED** [review]

File: Readme.md (hunk 6 above)
Comment (TimKolberger) [reply Readme.md]: I aligned this functionality with environment variables in a shell. 

```sh
VAR=123 # set
VAR=    # empty/remove
```

```sh
juli alias DAILY=JIRA-123  # set alias
juli alias DAILY=JIRA-456  # update alias
juli alias DAILY=          # remove alias
```

With this declarative API the user does not have the recall the state of an alias. The user describes the desired state of the alias - not what the user wants to do with the alias.
I think it delivers a better user experience in comparison to the imperative create, update and delete API:

```sh
juli alias create "DAILY" "JIRA-123" "DESCRIPTION" 
juli alias update "DAILY" "JIRA-456" "OTHER DESCRIPTION" 
juli alias delete "DAILY"
```

File: src/Parser.ts (hunk 11 above)
Comment (TimKolberger) [reply src/Parser.ts:63]: Without this check it recognises a "start time line" as ticket id.
```
#1.1.
12        <---- This "start time line"
  JIRA-123: some description
13
```

I found this location in the code to be only place to resolve the aliases because the validation of the time log entry is done while parsing.

File: src/index.ts (hunk 9 above)
Comment (TimKolberger) [reply src/index.ts:46]: With this format (without space) the output is consistent with the expected input. Is this intended behavior?

File: src/repository/AliasStore.ts (hunk 8 above)
Comment (TimKolberger) [reply src/repository/AliasStore.ts]: 👍

File: src/repository/AliasStore.ts (hunk 7 above)
Comment (TimKolberger) [reply src/repository/AliasStore.ts]: These are just typings - no curry today 🍛

File: src/repository/AliasStore.ts (hunk 10 above)
Comment (TimKolberger) [reply src/repository/AliasStore.ts]: No good arguments - just old habit.
I will update the path to not pollute the `jira` directory and refactor it to a const variable.

File: src/Parser.ts (hunk 11 above)
Comment (TimKolberger) [reply src/Parser.ts:63]: I would be happy to get your input on where we could resolve the aliases, without introducing breaking changes to the algorithm.

File: src/Parser.ts (hunk 5 above)
Comment (TimKolberger) [reply src/Parser.ts:8]: The function `parseTime` has no dependency on the Parser instance - there is no need to make it not static.

File: src/repository/AliasStore.ts (hunk 7 above)
Comment (TimKolberger) [reply src/repository/AliasStore.ts]: I extracted the function typing to avoid confusion

File: src/repository/AliasStore.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
+import * as fs from "fs-extra";
+import { CREATE_PATH } from "../const";
+import * as path from "path";
+import { InvalidArgumentError } from "../errors/InvalidArgumentError";
+import { IAlias, IAliases } from "../types";
+
+export const ALIAS_FILE = path.join("alias.json");
+
+export type AliasResolver = (idOrName: string) => IAlias;
```
Comment (Agreon) [inline src/repository/AliasStore.ts:9]: What do you use this type for? Couldn't you just type the method like:
```
  public resolveAlias = (idOrName: string): IAlias => {
```

**Review by Agreon — CHANGES_REQUESTED** [review]
